src/SizedArray.jl

```diff
@@ -53,43 +64,156 @@ end
+    return SizedArray{Tuple{S},T,1,1,Vector{T}}(x)
+end
+# disambiguation
+@inline function SizedVector{S}(a::StaticVector{S,T}) where {S,T}
+    return SizedVector{S,T}(a.data)
+end
 
-SizedMatrix{S1,S2,T,M} = SizedArray{Tuple{S1,S2},T,2,M}
-@inline SizedMatrix{S1,S2}(a::Array{T,M}) where {S1,S2,T,M} = SizedArray{Tuple{S1,S2},T,2,M}(a)
-@inline SizedMatrix{S1,S2}(x::NTuple{L,T}) where {S1,S2,T,L} = SizedArray{Tuple{S1,S2},T,2,2}(x)
+const SizedMatrix{S1,S2,T} = SizedArray{Tuple{S1,S2},T,2}
+
+@inline function SizedMatrix{S1,S2}(
+    a::TData,
+) where {S1,S2,T,M,TData<:AbstractArray{T,M}}
+    return SizedArray{Tuple{S1,S2},T,2,M,TData}(a)
+end
+@inline function SizedMatrix{S1,S2}(x::NTuple{L,T}) where {S1,S2,T,L}
+    return SizedArray{Tuple{S1,S2},T,2,2,Matrix{T}}(x)
+end
+@inline function SizedMatrix{S1,S2,T}(x::NTuple{L}) where {S1,S2,T,L}
+    return SizedArray{Tuple{S1,S2},T,2,2,Matrix{T}}(x)
+end
+# disambiguation
+@inline function SizedMatrix{S1,S2}(a::StaticMatrix{S1,S2,T}) where {S1,S2,T}
+    return SizedMatrix{S1,S2,T}(a.data)
+end
 
 Base.dataids(sa::SizedArray) = Base.dataids(sa.data)
 
-function (::Size{S})(a::Array) where {S}
-    Base.depwarn("`Size{S}(a::Array)` is deprecated, use `SizedVector{N}(a)`, `SizedMatrix{N,M}(a)` or `SizedArray{Tuple{S}}(a)` instead", :Size)
-    SizedArray{Tuple{S...}}(a)
+function promote_rule(
+    ::Type{SizedArray{S,T,N,M,TDataA}},
+    ::Type{SizedArray{S,U,N,M,TDataB}},
+) where {S,T,U,N,M,TDataA,TDataB}
+    TU = promote_type(T, U)
+    return SizedArray{S, TU, N, M, promote_type(TDataA, TDataB)}
+end
+
+function promote_rule(
+    ::Type{SizedArray{S,T,N,M}},
+    ::Type{SizedArray{S,U,N,M}},
+) where {S,T,U,N,M,}
+    TU = promote_type(T, U)
+    return SizedArray{S, TU, N, M}
 end
 
+function promote_rule(
+    ::Type{SizedArray{S,T,N}},
+    ::Type{SizedArray{S,U,N}},
+) where {S,T,U,N}
+    TU = promote_type(T, U)
+    return SizedArray{S, TU, N}
+end
+
+
+### Code that makes views of statically sized arrays also statically sized (where possible)
+
+@generated function new_out_size(::Type{Size}, inds...) where Size
+    os = []
+    map(Size.parameters, inds) do s, i
+        if i <: Integer
+            # dimension is fixed
+        elseif i <: StaticVector
+            push!(os, i.parameters[1].parameters[1])
+        elseif i == Colon || i <: Base.Slice
+            push!(os, s)
+        elseif i <: SOneTo
+            push!(os, i.parameters[1])
+        else
+            error("Unknown index type: $i")
+        end
+    end
+    return Tuple{os...}
+end
+
+function Base.view(
+    a::SizedArray{S},
+    indices::Union{Integer, Colon, StaticVector, Base.Slice, SOneTo}...,
+) where {S}
+    new_size = new_out_size(S, indices...)
+    return SizedArray{new_size}(view(a.data, indices...))
+end
 
-function promote_rule(::Type{<:SizedArray{S,T,N,M}}, ::Type{<:SizedArray{S,U,N,M}}) where {S,T,U,N,M}
-    SizedArray{S,promote_type(T,U),N,M}
+function Base.view(
```

Comment:
We really need to move this to the MArray.jl file.

Reply:
Sure, I'll move it.

Reply:
I think this can be marked as resolved?